Library/Homebrew/diagnostic.rb

```diff
@@ -341,6 +341,19 @@ def check_tmpdir_sticky_bit
       end
       alias generic_check_tmpdir_sticky_bit check_tmpdir_sticky_bit
 
+      def check_xdg_data_dirs
+        xdg_data_dirs = ENV.fetch("XDG_DATA_DIRS", nil)
+        return if xdg_data_dirs.blank? || xdg_data_dirs.split("/").include?(HOMEBREW_PREFIX/"share")
+
+        <<~EOS
+          Homebrew's share was not found in your XDG_DATA_DIRS but you have
+          this variable set to include other locations.
+          Some programs like `vapigen` may not work correctly.
+          Consider adding Homebrew's share directory to XDG_DATA_DIRS like so:
+            echo 'export XDG_DATA_DIRS="#{HOMEBREW_PREFIX}/share:$XDG_DATA_DIRS"' >> #{Utils::Shell.profile}
```

Comment:
> Telling the user add something to the shell profile alongside `brew shellenv` but not wanting to put it in `brew shellenv` feels very weird to me.

Agreed.

> If there's strong reasons to not have it in `brew shellenv`, why do those reasons not apply here? Why fail `brew doctor` and unconditionally tell everyone to do something we are not doing ourselves? The messaging after all is implying this is a very good idea to set. Or are we saying the message here is also wrong?

Might be some answers from discussion in the previous PR [here](https://github.com/Homebrew/brew/pull/17932#discussion_r1700662345), but, to me:
- I don't think there are strong reasons to not have it in `brew shellenv`.
- We should probably just do it ourselves in `shellenv` in addition to doing it in `brew doctor` for the reason that @gromgit outlines [here](https://github.com/Homebrew/brew/pull/17932#issuecomment-2266343270).
- I don't think the message here is wrong.

And to address some points from the previous discussion:

#### We don't want XDG on macOS

I don't really see why. XDG is a fine Unix standard (it's used on BSD too, and not just Linux). In any case, I think the ship has sailed on keeping it out. There are already two dozen formulae in Homebrew/core that explicitly reference `XDG_*`, and probably many more that use it with no reference in the formula.

#### macOS users don't set `XDG_*`, so we don't need this

Maybe that's true, but, even so, I don't think doing this hurts. We already need it on Linux anyway. I don't think having this code is such a burden that we ought instead to wait and see if a macOS user gets tripped up by this before adding it. Doing it everywhere is simpler and easier to maintain than just doing it on Linux too.

Reply:
> Telling the user add something to the shell profile alongside `brew shellenv` but not wanting to put it in `brew shellenv` feels very weird to me.

There's a multiple reasons for this:
- this is not as widely applicable as other variables in `brew shellenv`. In before someone says "no-one uses `INFOPATH`": I would be in favour of removing that too, on that basis
- I do not want to keep adding to `brew shellenv` based on very small numbers of issues
- I do not want to be fixing `brew shellenv` issues on macOS when they seem only applicable to Linux
- I want to keep `brew shellenv` output as short, minimal and human-readable as possible
- I want to keep `brew shellenv` as performant as possible
- I have had a bunch of issues in the past where people have been trying moderately hard to say "we should make macOS follow the XDG spec", we have decided not to do so, and doing this would weaken this (frankly tiresome at this point) argument

And, just because this seems to be one of the rare cases where I have to regrettable do this:
- I've been maintaining this project for 15 years, being a "product manager" and "project leader" for it for a while and have learned a few things along the way. Along this way (supporting/showing `XDG_*` variables on macOS) only leads us to pain and additional code to support on our majority platform based on a tiny, tiny number of people who actually care about this. Arguing about these sort of minutiae is not a good use of time for anyone on this thread and I'm going to ask nicely if we can please move on from this.

> * I don't think there are strong reasons to not have it in `brew shellenv`.

@carlocab I'm afraid I read from this "Mike has reasons, I just don't like them". I find it pretty rude, honestly, that I've taken time on multiple occasions to explain this only to have you dismiss my reasoning.

> I don't really see why. XDG is a fine Unix standard (it's used on BSD too, and not just Linux). In any case, I think the ship has sailed on keeping it out. There are already two dozen formulae in Homebrew/core that explicitly reference `XDG_*`, and probably many more that use it with no reference in the formula.

It's a "free desktop" standard. macOS is not a "free desktop". It is a different thing. Homebrew is primarily a macOS package manager. There have been similar arguments/reasons made about why we don't follow the FHS "standard" similarly. We are primarily macOS software. We do not follow standards just because they exist. In this case: they do not make sense for us.

I've said before and I'll say again: the more support for `XDG_*` we put into Homebrew/brew: the more we're going to be asked to support it for all cache, log, config etc. files and, if we support that, the more we're going to be asked to make it the default paths. You can (and probably will) say "this will not happen" and I will disagree with that. I would, as I've said above, gently nudge you with "I have more experience than anyone else on this thread on seeing what happens based on these sorts of changes in Homebrew".

> Maybe that's true, but, even so, I don't think doing this hurts.

Hopefully I've explained, at length, why I think this hurts.